### DISCUSSION TITLE
Where are validators / ValidationBehaviors?

### DISCUSSION THREAD

**Discussion by sinsedrix** [CommunityToolkit/Maui#169] [Q&A] [answered]
Will the validators in Xamarin CommunityToolkit ported to Maui CommunityToolkit?
Or is there a new pattern to validate inputs?

https://github.com/xamarin/XamarinCommunityToolkit/tree/main/src/CommunityToolkit/Xamarin.CommunityToolkit/Behaviors/Validators

**Comment by TheCodeTraveler** [comment, marked as answer]
Yes the proposals for the Validator Behaviors from Xamarin.Community Toolkit have been [approved for .NET MAUI Toolkit](https://github.com/CommunityToolkit/Maui/projects/1#column-16085863). 

You can follow our progress in this PR: https://github.com/CommunityToolkit/Maui/pull/142

**Comment by emorell96** [comment]
I was playing around with the validators (great job btw) and I was wondering why the ``ValidateAsync`` isn't a public function or at least ``ValidateNestedAsync``? It would be useful to be able to call it from outside the class specially for doing more specialized ``MultiValidationBehavior``.

I found this limitation when I tried to create a derived class from ``MultiValidationBehavior`` and override its ``ValidateAsync``. But outside this library, I cannot call the function you use in the ``MultiValidationBehavior`` since it's internal: ``ValidateNestedAsync``.

**Reply by VladislavAntonyuk** [reply]
To run the validation manually, you can call ```public ValueTask ForceValidate()```
It internally calls `ValidateAsync` and set all dependent properties like `IsRunning`

**Reply by emorell96** [reply]
True but does it block the thread since it's a sync function?

**Reply by VladislavAntonyuk** [reply]
it's ValueTask.
you can await it

**Reply by emorell96** [reply]
It is ValueTask never mind. But it is not cancellable though. That I'd say it's the advantage of the one you use internally. Specially for validation that needs to hit a web server.

**Reply by VladislavAntonyuk** [reply]
in that case, I suggest you open a proposal/disussion to set the CancellationToken parameter to the Task functions

**Reply by emorell96** [reply]
I will. Also can I ask, how would you measure performance in a UI library like this. I know how to benchmark with a console app, or in general background tasks, but how do you benchmark something that depends on the UI? I wanted to test if using Parallel.ForEachAsync instead of Task.WhenAll would have any measurable performance difference.

**Reply by VladislavAntonyuk** [reply]
I didn't do that, but I suppose you can do something similar which we have in our Unit Test library, where we call different UI controls by mocking some Maui services. @brminnick can provide more details about benchmarks

**Reply by emorell96** [reply]
I'd appreciate those details, thank you. I already have a Parallel.ForEachAsync version that I'd like to test vs the current implementation and see if it does any better out of curiosity.

I've opened the discussion (https://github.com/CommunityToolkit/Maui/discussions/225) about adding the Token. And I'll be happy to contribute with the changes it requires if approved :)

**Comment by CliffAgius** [comment]
Closed as merged.
